### PR TITLE
arm/armv8-r: fix unable to switch context in ISR context

### PR DESCRIPTION
**Pull request by anchao**
## Summary

arm/armv8-r: fix unable to switch context in ISR context

Regression by https://github.com/apache/nuttx/pull/13651:
```
| commit 35c8c80a00a99ff0e19d51eaa74165bd830a36f8
| Author: ligd <liguiding1@xiaomi.com>
| Date:   Fri Jul 26 23:14:13 2024 +0800
|
| arch: change nxsched_suspend/resume_scheduler() called position
```

Signed-off-by: chao an <anchao@lixiang.com>

## Impact

N/A

## Testing

CI-CHECK

### CODE REVIEW

File: arch/arm/src/armv8-r/arm_doirq.c [hunk 1]
```diff
@@ -58,7 +58,9 @@ uint32_t *arm_doirq(int irq, uint32_t *regs)
    * and we will use this_task(). Therefore, it cannot be overridden.
    */
 
+#ifdef CONFIG_SMP
```
Comment (GUIDINGLI) [inline arch/arm/src/armv8-r/arm_doirq.c:61]: split to another PR

Comment (anchao) [reply arch/arm/src/armv8-r/arm_doirq.c:61]: no, this is regression of arm_doirq()

Comment (GUIDINGLI) [reply arch/arm/src/armv8-r/arm_doirq.c:61]: Why not, this not the introduce by https://github.com/apache/nuttx/pull/13651/files#diff-723f44e76d5ebc519459ee93d6d0e42a7cea30ad0343b61e3ab48a491f2aaf58

**Review by GUIDINGLI — CHANGES_REQUESTED** [review]
The SMP is another questions, so you should split to another patch or PR

**Comment by anchao** [comment]
> The SMP is another questions, so you should split to another patch or PR

You should check your submission carefully if which have any rookie mistakes.

**Comment by anchao** [comment]
@GUIDINGLI @hujun260 

After your and hujun's changes:
1. any irq_dispatch() call needs to properly refresh tcb to check if context switch has occurred
2. The trace location also needs to be updated carefully
3. new current registers must be update and return from ISR.

but your previous changes completely missed on some arm-specific chips:


image

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by GUIDINGLI** [comment]
@anchao 
Why NOT two patches fix two questions ?

**Comment by anchao** [comment]
> @anchao Why NOT two patches fix two questions ?

@GUIDINGLI  https://github.com/apache/nuttx/issues/14514 :)
